compiler/resolution/lowerIterators.cpp

```diff
@@ -1528,13 +1530,77 @@ static void markLoopProperties(ForLoop* forLoop, BlockStmt* ibody,
   }
 }
 
+// This function is called while expanding an iterator for a for/foreach loop
+// that may potentially have intents / shadow variables.
+// While processing these update a symbol map that we'll use to map shadow
+// variables to whatever they should be mapped to in the expanded loop.
+static void processShadowVariables(ForLoop* forLoop, SymbolMap *map) {
+  for_shadow_vars (svar, temp, forLoop) {
+    switch (svar->intent) {
+      case TFI_DEFAULT:
+      case TFI_CONST:
+        INT_ASSERT(false);
+
+      case TFI_IN:
+        {
+        // If we have a variable with an 'in' intent for a foreach loop we'll
+        // want to capture its value before we start executing the loop. We'll
+        // use this copied version to give an initial value to thread-private
+        // versions of the variable.  How many thread private variables we
+        // should have and how to initialize them is something handled by later
+        // passes since at this point we don't know if this loop will be
+        // vectorized or gpuized, so in the meantime we model this as an
+        // assignment to a primitive and leave it to future passes (like
+        // gpuTransforms) to rewrite things as appropriate.
+        // IOW: coming out of this case we'll add something that looks like this:
+        //
+        //   var capX = x;
+        //   var taskIndX = PRIM_TASK_IND_CAPTURE_OF(capX);
+        //   # note: there's also a flag on taskIndX marking it as being a "task"-independent variable
+
+        SET_LINENO(forLoop);
+
+        VarSymbol* capturedSvar = new VarSymbol(svar->name, svar->type);
+        forLoop->insertBefore(new DefExpr(capturedSvar));
+        forLoop->insertBefore(new CallExpr(PRIM_MOVE, capturedSvar, svar->outerVarSE->symbol()));
+
+        VarSymbol* taskIndVar = new VarSymbol(svar->name, svar->type);
+        taskIndVar->addFlag(FLAG_TASK_PRIVATE_VARIABLE);
+        forLoop->insertBefore(new DefExpr(taskIndVar));
+        forLoop->insertBefore(new CallExpr(PRIM_ASSIGN, taskIndVar, new CallExpr(PRIM_TASK_INDEPENDENT_SVAR_CAPTURE, capturedSvar)));
+
+        map->put(svar, taskIndVar);
+        }
+        break;
+
+      case TFI_CONST_IN:
+      case TFI_REF:
+      case TFI_CONST_REF:
+      case TFI_REDUCE_OP:
+        INT_ASSERT(false);
```

Comment:
This, and the assertion below -- they are already being errored out by the parser, correct? At least consider adding a comment for the person who might be debugging these assertions.

Reply:
"REDUCE_OP" is. I'm not so sure about "CONST_REF" and "REF", either way if we change the parser there'd be work to be done here so I definitely makes sense to assert out.

I've added a "// to be implemented" comment.